posthog/migrations/0243_unpack_plugin_source_files.py

```diff
@@ -0,0 +1,90 @@
+    # Only 4 changes:
+    # - Plugin and PluginSourceFiles have been stripped from types
+    #   (they have to be vars in this scope, but vars cannot be used as types)
+    # - plugin_json cannot be provided as an arg
+    # - records are `create()`d instead of `update_or_create()`d
+    # - there's no return value
+    def update_or_create_from_plugin_archive(plugin):
+        """Create PluginSourceFile objects from a plugin that has an archive."""
+        if plugin.archive is None:
+            raise exceptions.ValidationError(
+                f"Could not extract files from plugin {plugin.name} ID {plugin.id} - it has no archive"
+            )
+        # Extract plugin.json - required, can be provided to the function as an optimization
+        plugin_json = cast(Optional[Dict[str, Any]], get_file_from_archive(plugin.archive, "plugin.json"))
+        if not plugin_json:
+            raise exceptions.ValidationError(f"Could not find plugin.json in plugin {plugin.name} ID {plugin.id}")
+        # Extract frontend.tsx - optional
+        frontend_tsx: Optional[str] = get_file_from_archive(plugin.archive, "frontend.tsx", json_parse=False)
+        # Extract index.ts - optional if frontend.tsx is present, otherwise required
+        main_filename_defined = plugin_json.get("main")
+        main_filenames_to_try = [main_filename_defined] if main_filename_defined else ["index.js", "index.ts"]
+        index_ts: Optional[str] = None
+        for main_filename in main_filenames_to_try:
+            if index_ts := get_file_from_archive(plugin.archive, main_filename, json_parse=False):
+                break
+        else:
+            if frontend_tsx is None:
+                raise exceptions.ValidationError(
+                    f"Could not find main file {' or '.join(main_filenames_to_try)} in plugin {plugin.name} ID {plugin.id}"
+                )
+        # Save plugin.json
+        PluginSourceFile.objects.create(plugin=plugin, filename="plugin.json", source=json.dumps(plugin_json))
+        # Save frontend.tsx
+        if frontend_tsx:
+            PluginSourceFile.objects.create(plugin=plugin, filename="frontend.tsx", source=frontend_tsx)
+        # Save index.ts
+        if index_ts:
+            # The original name of the file is not preserved, but this greatly simplifies the rest of the code,
+            # and we don't need to model the whole filesystem (at this point)
+            PluginSourceFile.objects.create(plugin=plugin, filename="index.ts", source=index_ts)
+
+    # Source plugins have already been migrated in 0233_plugin_source_file, while local ones don't store code in the DB
+    for plugin in Plugin.objects.exclude(plugin_type="source").exclude(plugin_type="local"):
+        try:
+            update_or_create_from_plugin_archive(plugin)
+        except exceptions.ValidationError as e:
+            logger.warn(f"Migration 0243 - failed to extract or save code of plugin {plugin.name} ID {plugin.id}: {e}")
```

Comment:
Unclear: Is this actionable somehow? What is the user supposed to do.

Reply:
This is just a warning – in some unlikely case the archive might be broken (though the user should know about this already, as such a plugin could not be enabled). Used "skipping" wording to indicate that this is not an error that MUST be taken care of.

Reply:
I can see a user upgrading and getting confused or asking about this warning. Do you think we should reword to minimize that pain?

Reply:
This should never really happen, so I think the message is fine after the update – the user probably _should_ message us if they see it, as this either means their custom plugin stored in GitHub/npm is malformed (and if they don't know about this, they'll have a bad time anyway) or an approved repository plugin is malformed (but I just ran the migration with the majority of repository plugins and found no problems).